igvc_navigation/src/field_d_planner/main.cpp

```diff
@@ -0,0 +1,268 @@
+  expanded_cloud.clear();
+  expanded_cloud.header.frame_id = "odom";
+
+  for (std::tuple<int, int> ind : inds)
+  {
+    pcl::PointXYZRGB p;
+
+    p.x = static_cast<float>(std::get<0>(ind) - x_initial) * map->resolution;
+    p.y = static_cast<float>(std::get<1>(ind) - y_initial) * map->resolution;
+    p.z = -0.05f;
+
+    // set the node color to red if its g-value is inf. Otherwise blue.
+    if (planner.getG(Node(ind)) == std::numeric_limits<float>::infinity())
+    {
+      uint8_t r = 255, g = 0, b = 0;
+      uint32_t rgb = ((uint32_t)r << 16 | (uint32_t)g << 8 | (uint32_t)b);
+      p.rgb = *reinterpret_cast<float*>(&rgb);
+    }
+    else
+    {
+      uint8_t r = 0, g = 125, b = 125;
+      uint32_t rgb = ((uint32_t)r << 16 | (uint32_t)g << 8 | (uint32_t)b);
+      p.rgb = *reinterpret_cast<float*>(&rgb);
+    }
+
+    expanded_cloud.points.push_back(p);
+  }
+
+  expanded_pub.publish(expanded_cloud);
+}
+
+//--------------------------- ROS Callbacks ----------------------------//
+
+/**
+    Set the current map to be used by the D* Lite search problem. The initial
+    map is used to perform the first search through the occupancy grid (equivalent
+    to A*). All maps thereafter are used to update edge costs for the search problem.
+*/
+void map_callback(const igvc_msgs::mapConstPtr& msg)
+{
+  std::lock_guard<std::mutex> planning_lock(planning_mutex);
+  map = msg;  // update current map
+
+  if (initialize_graph)
+  {
+    // initial x and y coordinates of the graph search problem
+    x_initial = static_cast<int>(map->x_initial);
+    y_initial = static_cast<int>(map->y_initial);
+    planner.NodeGrid.initializeGraph(map);
+    initialize_graph = false;
+  }
+}
+
+/**
+    Assigns a valid goal to the graph search problem. Goal index obtained by
+    converting from the /map frame goal coordinate to the graph index.
+
+    @param[in] msg the message received on the "/waypoint" topic
+*/
+void waypoint_callback(const geometry_msgs::PointStampedConstPtr& msg)
+{
+  std::lock_guard<std::mutex> lock(planning_mutex);
+
+  int goal_x, goal_y;
+  goal_x = static_cast<int>(std::round(msg->point.x / planner.NodeGrid.Resolution)) + x_initial;
+  goal_y = static_cast<int>(std::round(msg->point.y / planner.NodeGrid.Resolution)) + y_initial;
+
+  std::tuple<int, int> new_goal = std::make_tuple(goal_x, goal_y);
+
+  // re-initialize graph search problem if goal has changed
+  if (planner.NodeGrid.Goal.getIndex() != new_goal)
+    goal_changed = true;
+
+  planner.NodeGrid.setGoal(new_goal);
+
+  float distance_to_goal = planner.NodeGrid.euclidianHeuristic(new_goal) * planner.NodeGrid.Resolution;
+
+  ROS_INFO_STREAM((goal_changed ? "New" : "Same") << " waypoint received. Search Problem Goal = " << goal_x << ", "
+                                                  << goal_y << ". Distance: " << distance_to_goal << "m.");
+
+  if (distance_to_goal > maximum_distance)
+  {
+    ROS_WARN_STREAM("Planning to waypoint more than " << maximum_distance
```

Comment:
Change to `ROS_WARN_STREAM_THROTTLE` so the terminal isn't spammed

Reply:
Done :+1: